### PR TITLE
Fix xml navigation on xml union type

### LINKED ISSUE #28692: Enable xml navigation access on xml union types
**Description:**
As of now, xml navigation access on xml unions give a compilation error with fixes in #28556. This needs to be enabled. 

**Steps to reproduce:**
```
'xml:ProcessingInstruction x1 = xml `<?xml-stylesheet href="mystyle.css" type="text/css"?>`;
xml x10 = x1[0]/*;

xml<'xml:Element> | xml<'xml:Comment> x2 = xml `<root><!-- comment node--><name>anne</name><city>colombo</city></root>`;
xml x10 = x2/*;
```

### PR DESCRIPTION
**Pull request by poorna2152**
## Purpose
Fixes #28692 

## Approach
Remove the check for union type in the TypeChecker for the xml navigation expression, because union type is allowed here in the `checkExpr` call for the `expr` in the navigationExpr guarantee that the expr is of the subtype `xml`.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml-navigation-access.bal [hunk 1]
```diff
@@ -170,6 +170,34 @@ function testXMLNavigationWithEscapeCharacter() {
...
+    assert(x3/*.<b>, xml ``);
+    assert(x3/**/<c>, xml ``);
+    assert(x3/<d>, xml ``);
+
+    xml<'xml:Text>|xml<'xml:ProcessingInstruction> x4 = xml `test xml text`;
+    assert(x4/*, xml ``);
+    assert(x4/*.<d>, xml ``);
+    assert(x4/**/<e>, xml ``);
+    assert(x4/<f>, xml ``);
+}
```
Comment (SasinduDilshara) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml-navigation-access.bal:199]: shall we add a test type reference types.

```
type A xml:Comment;

type B xml:Element;

type C A|B;

```

**Review by LakshanWeerasinghe — COMMENTED** [review]
@MaryamZi Do we need to remove diagnostic from [DiagnosticErrorCode](https://github.com/ballerina-platform/ballerina-lang/blob/b30581db0187b133af6a0d874c199441de98e40c/compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java#L732) and [compiler.properties](https://github.com/LakshanWeerasinghe/ballerina-lang/blob/b30581db0187b133af6a0d874c199441de98e40c/compiler/ballerina-lang/src/main/resources/compiler.properties#L712) files?

**Comment by LakshanWeerasinghe** [comment]
The [spec](https://ballerina.io/spec/lang/master/#section_6.36) says.
> XML navigation expressions allow for convenient navigation of XML element structure, in a similar way to XPath.

One way to interpret this is that XML navigation is supported only for XML values with the `xml:Element` subtype. In a scenario where an XML value has the type `xml:Comment|xml:Text`, it seems okay to issue a compile-time error, as users wouldn't be able to find anything useful from such an operation. wdyt? @MaryamZi

**Comment by poorna2152** [comment]
Was not able to add a test with an xml element sequence due to this [bug](https://github.com/ballerina-platform/ballerina-lang/issues/42364)

**Comment by MaryamZi** [comment]
> @MaryamZi Do we need to remove diagnostic from [DiagnosticErrorCode](https://github.com/ballerina-platform/ballerina-lang/blob/b30581db0187b133af6a0d874c199441de98e40c/compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java#L732) and [compiler.properties](https://github.com/LakshanWeerasinghe/ballerina-lang/blob/b30581db0187b133af6a0d874c199441de98e40c/compiler/ballerina-lang/src/main/resources/compiler.properties#L712) files?

Yeah, we can remove. Just need to leave the rest of the diagnostic codes (and IDs) as they are.

**Comment by MaryamZi** [comment]
> The [spec](https://ballerina.io/spec/lang/master/#section_6.36) says.
> 
> > XML navigation expressions allow for convenient navigation of XML element structure, in a similar way to XPath.
> 
> One way to interpret this is that XML navigation is supported only for XML values with the `xml:Element` subtype. 

If we do this, we'll be disallowing access on `xml` also, won't we? 

> In a scenario where an XML value has the type `xml:Comment|xml:Text`, it seems okay to issue a compile-time error, as users wouldn't be able to find anything useful from such an operation. wdyt? @MaryamZi

The spec says for the expression, "The static type of the expression must be a subtype of xml.".

So from a spec point of view this is OK. Also when we perform the access on non-element XML values, we get an empty sequence rather than an error, so if at all, I think we can give a warning to say it will always be an empty sequence (I don't think we need to), but we shouldn't be giving an error.

**Comment by MaryamZi** [comment]
> Was not able to add a test with an xml element sequence due to this [bug](https://github.com/ballerina-platform/ballerina-lang/issues/42364)

Can't we work around it with something like 

```ballerina
xml<xml:Element>|xml:Comment x = <xml<xml:Element>> xml `<a></a><b></b>`;
```

?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/xml/XMLAccessTest.java [hunk 2]
```diff
@@ -269,6 +269,11 @@ public void testXmlAccessWithLargerIndex() {
         BRunUtil.invoke(result, "testXmlAccessWithLargerIndex");
     }
 
+    @Test
+    public void testXMLNavigationWithUnionType() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/xml/XMLAccessTest.java:273]:
```suggestion
    public void testXmlNavigationWithUnionType() {
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/xml/XMLAccessTest.java [hunk 3]
```diff
@@ -269,6 +269,11 @@ public void testXmlAccessWithLargerIndex() {
...
+        BRunUtil.invoke(navigation, "testXMLNavigationWithUnionType");
+    }
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/xml/XMLAccessTest.java:275]: Please also add some negative tests against access on unions (with and without xml members) that as a whole are not subtypes of xml.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/xml/XMLAccessTest.java (hunk 2 above)
Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/xml/XMLAccessTest.java:273]: Fix the test function also.

**Review by MaryamZi — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml_step_expr_negative.bal [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    _ = x4/*;
+    _ = x4/*.<d>;
+
+    XCE x5 = xml `<foo><bar>b</bar></foo>`;
+    _ = x5/*.<bar>;
+    _ = x5/**/<bar>;
+
+    xml<xml:Element>|xml<xml:Comment>|xml<xml:ProcessingInstruction>|xml:Text|int x6 = xml `<l><m><n></n></m></l>`;
+    _ = x6/*;
+    _ = x6/*.<m>;
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml_step_expr_negative.bal:44]: Shall we add more tests for something like `x6/*.<m|n>` (check for element m or n) also `x6/*.<m>[1] `(if we have multiple <m>)

**Pull request merged**